File: product-consumer-service/src/main/java/de/hdi/product_consumer_service/config/GrpcClientConfig.java
package de.hdi.product_consumer_service.config;

import java.util.concurrent.TimeUnit;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import de.hdi.product_producer_service.ProductServiceGrpc;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

@Configuration
public class GrpcClientConfig {

    @Value("${spring.grpc.client.product-producer.address}")
    private String productServiceAddress;

    @Bean
    public ManagedChannel productServiceChannel() {
        String[] parts = productServiceAddress.split(":");
        String host = parts[0];
        int port = 9090; // Default port

        if (parts.length > 1) {
            try {
                port = Integer.parseInt(parts[1]);
            } catch (NumberFormatException e) {
                // Use default port if parsing fails
            }
        }

        return ManagedChannelBuilder.forAddress(host, port)
                .usePlaintext()
                .keepAliveTime(30, TimeUnit.SECONDS)
                .build();
    }

    @Bean
    public ProductServiceGrpc.ProductServiceBlockingStub productServiceBlockingStub(ManagedChannel productServiceChannel) {
        return ProductServiceGrpc.newBlockingStub(productServiceChannel);
    }
}

File: order-kotlin-service/src/main/kotlin/de/hdi/order_kotlin_service/grpc/config/GrpcClientConfig.kt
package de.hdi.order_kotlin_service.config

import de.hdi.product_producer_service.ProductServiceGrpc
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class GrpcClientConfig {

    private val logger = LoggerFactory.getLogger(GrpcClientConfig::class.java)

    @Value("\${spring.grpc.client.product-producer.address:localhost:9090}")
    private lateinit var productServiceAddress: String

    @Bean
    fun productServiceChannel(): ManagedChannel {
        val host: String
        val port: Int
        if (productServiceAddress.contains(":")) {
            val parts = productServiceAddress.split(":")
            host = parts[0]
            port = parts[1].toInt()
        } else {
            host = productServiceAddress
            port = 9090 // Default gRPC port
        }

        System.setProperty("io.netty.transport.noNative", "true")
        System.setProperty("io.grpc.ChannelImpl.disableDetectingTransport", "true")

        return ManagedChannelBuilder.forAddress(host, port)
            .usePlaintext()
            .build()
    }

    @Bean
    fun productServiceStub(productServiceChannel: ManagedChannel): ProductServiceGrpc.ProductServiceBlockingStub {
        logger.info("Creating ProductServiceBlockingStub")
        return ProductServiceGrpc.newBlockingStub(productServiceChannel)
    }
}